### PR TITLE
Add event/audit logging on all FHIR mutations

### PR DESCRIPTION
**Pull request by pbugni**
All mutating calls routed through `jwt-proxy` will now generate log entries.  When configured to use JSON formatting (also part of this PR) and the environment (i.e. https://github.com/uwcirg/helloworld-environments/pull/4) is configured with a logserver, such log entries will land in the log server.

Example format for a new Patient PUT:
```json
  {
    "id": 101,
    "event": {
      "name": "jwt_proxy_event_logger",
      "tags": [
        "Patient",
        "PUT"
      ],
      "user": "test",
      "params": {
        "given": "Liz",
        "family": "Taylor",
        "birthdate": "1999-01-01"
      },
      "asctime": "2023-01-24 01:23:00,956",
      "message": "PUT Patient",
      "levelname": "INFO",
      "event_version": "1"
    }
  }
```

and for DELETE of the same patient:
```json
  {
    "id": 103,
    "event": {
      "name": "jwt_proxy_event_logger",
      "tags": [
        "Patient",
        "DELETE"
      ],
      "user": "test",
      "asctime": "2023-01-24 01:23:48,836",
      "message": "DELETE Patient/61",
      "subject": "Patient/61",
      "levelname": "INFO",
      "event_version": "1"
    }
  }
```

### CODE REVIEW

**Comment by mcjustin** [comment]
Thanks @pbugni ! I'm confused with your example log entries tho:

1. Why does the PUT example have references to anything about DELETE?
2. The PUT has `version` (app version, I think) but not `event_version` (version of the event spec, I think); visa-versa for the DELETE example.
3. The DELETE doesn't indicate what type of resource is being deleted.

**Comment by pbugni** [comment]
> Thanks @pbugni ! I'm confused with your example log entries tho:
> 
>     1. Why does the PUT example have references to anything about DELETE?
> 
>     2. The PUT has `version` (app version, I think) but not `event_version` (version of the event spec, I think); visa-versa for the DELETE example.
> 
>     3. The DELETE doesn't indicate what type of resource is being deleted.

Bah!  Sorry @mcjustin - sloppy cut & paste.  I corrected the comment to show the jwt_proxy PUT example as generated.  Also corrected parsing for ResourceType on DELETE.

**Review by mcjustin — APPROVED** [review]

File: jwt_proxy/api.py [hunk 1]
```diff
@@ -54,6 +70,7 @@ def validate_jwt(relative_path):
     response_content = proxy_request(
         req=request,
         upstream_url=f"{current_app.config['UPSTREAM_SERVER']}/{relative_path}",
+        user_info=decoded_token.get("preferred_username"),
```
Comment (ivan-c) [inline jwt_proxy/api.py:73]: Email should be slightly more consistent than Keycloak username
```suggestion
        user_info=decoded_token.get("email"),
```

File: jwt_proxy/audit.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+def deets_from_url(url, resource_type, id):
+    """Given url, and best guess at resource_type and id, return best guess"""
+    if resource_type and id:
+        return resource_type, id
+
+    # url: UPSTREAM_SERVER/fhir/ResourceType/<id or params>
+    parsed = urlparse(url)
+    if not parsed.path.startswith("/fhir/"):
+        audit_entry(f"Unexpected fhir path: {url} can't parse", level="error")
```
Comment (ivan-c) [inline jwt_proxy/audit.py:46]: We may need to make this configurable later. Outside of fEMR, the FHIR base URL may be inconsistent (eg used to be `/hapi-fhir-jpaserver/fhir`)

**Review by ivan-c — APPROVED** [review]
Looks good!!

File: jwt_proxy/api.py (hunk 1 above)
Comment (pbugni) [reply jwt_proxy/api.py]: i didn't find "email" present in testing, could just be our test user lacks that attribute.
just pushed a change to prefer "email" if defined, or fall back to "preferred_username"

File: jwt_proxy/audit.py (hunk 2 above)
Comment (pbugni) [reply jwt_proxy/audit.py:46]: Yes, that was the point of logging the error, should we fail to match this expectation.

**Pull request merged**